datastore/adapter/keytransform.py

```diff
@@ -290,3 +295,161 @@ class BinaryNestedPathAdapter(_NestedPathAdapter[datastore.abc.BinaryDatastore],
 
 class ObjectNestedPathAdapter(_NestedPathAdapter[datastore.abc.ObjectDatastore], datastore.abc.ObjectAdapter):  # noqa: E501
 	...
+
+
+class _FlatFSAdapter(_Adapter[DS], typing.Generic[DS]):
+	"""Represents a simple DatastoreAdapter that shards/namespaces incoming keys.
+
+	Incoming keys are sharded into namespaces according to specifications
+	in a `SHARDING` file, which `prefix`, `suffix`, and `next-to-last`
+	sharding functions available.
+
+	The file should be be a single line in the following format:
+		{prefix}/version/function_name/key_length
+	
+	Implementation Notes:
+		* The `default_prefix` is "/repo/flatfs/shard/"
+		* The only version supported is "v1"
+		* function_name must be `prefix`, `suffix`, or `next-to-last`
+
+	For example:
+
+		>>> import datastore
+		>>>
+		>>> ds = datastore.DictDatastore()
+		>>> np = datastore.FlatFSDatastore(ds)
+		>>>
+		>>> np.put(datastore.Key('/abcdefghijk'), 1)
+		>>> np.get(datastore.Key('/abcdefghijk'))
+		1
+		>>> ds.get(datastore.Key('/abcdefghijk'))
+		None
+		>>> ds.get(datastore.Key('/ab/cd/ef/abcdefghijk'))
+		1
+		>>> np.put(datastore.Key('abc'), 2)
+		>>> np.get(datastore.Key('abc'))
+		2
+		>>> ds.get(datastore.Key('/ab/ca/bc/abc'))
+		2
+	"""
+
+	_default_sharding_file: str = "SHARDING"
+	_default_prefix: str = "/repo/flatfs/shard/"
+	
+	@staticmethod
+	def _prefix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).ljust(length, "_")[:length])
+	
+	@staticmethod
+	def _suffix(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length, "_")[:length])
+
+	@staticmethod
+	def _next_to_last(key: datastore.Key, length: int) -> datastore.Key:
+		return datastore.Key(str(key).rjust(length + 1, "_")[:length])
+
+	def __init__(self, *args,
+	             sharding_file: str = None,
```

Comment:
`sharding_file` is a datastore key, not an actual file path. Change this (and all related properties) to `sharding_key: datastore.Key` instead.